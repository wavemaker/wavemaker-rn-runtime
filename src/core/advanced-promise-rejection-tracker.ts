/**
 * Advanced Promise Rejection Tracker
 *
 * Provides comprehensive promise rejection tracking across different JavaScript engines.
 * This module addresses the critical gap where API call failures and async function errors
 * are not caught by React Error Boundaries, ensuring all promise rejections are handled.
 *
 * Engine Support:
 * - Web: Uses browser's 'unhandledrejection' event
 * - Hermes: Uses HermesInternal.enablePromiseRejectionTracker
 * - JSC: Uses React Native's 'promise/setimmediate/rejection-tracking' module
 *
 * Coverage Includes:
 * - API call failures (fetch, axios, any HTTP library)
 * - Async function errors (unhandled throws in async/await)
 * - Manual Promise.reject() calls
 * - Third-party library promise rejections
 *
 * Research Notes:
 * This implementation was developed through extensive research into React Native's
 * promise handling across different engines. Each engine has its own optimal approach:
 * - Hermes provides a native tracking API that hooks into the engine's internals
 * - JSC can use React Native's internal promise tracking module
 * - Web browsers have standardized unhandledrejection events
 *
 * The approach automatically detects the best available method and falls back
 * gracefully when native tracking is unavailable.
 */

import { Platform } from 'react-native';

// TypeScript declarations for global objects
declare global {
  interface Global {
    HermesInternal?: {
      enablePromiseRejectionTracker?: (options: {
        allRejections: boolean;
        onUnhandled: (id: number, error: any) => void;
        onHandled: (id: number) => void;
      }) => void;
    };
  }
}

export interface PromiseRejectionHandler {
  (error: any): void;
}

/**
 * Multi-engine promise rejection tracking implementation
 */
export class AdvancedPromiseRejectionTracker {
  private onRejectionHandler: PromiseRejectionHandler;
  private cleanupFunctions: (() => void)[] = [];

  constructor(onRejection: PromiseRejectionHandler) {
    this.onRejectionHandler = onRejection;
  }

  /**
   * Setup promise rejection tracking based on available APIs
   * Returns true if native tracking was enabled, false if fallback used
   */
  public setup(): boolean {
    // Automatically detect and setup the best available promise rejection tracking method

    if (Platform.OS === 'web') {
      return this.setupWebTracking();
    }

    // For mobile platforms, try native tracking methods in order of preference
    if (this.setupHermesTracking()) {
      return true;
    }

    // Fallback to React Native internal tracking for JSC
    if (this.setupReactNativeTracking()) {
      return true;
    }

    // No native tracking available - this should rarely happen in modern RN versions
    return false;
  }

  /**
   * Method 1: Web Environment - Use unhandledrejection event
   */
  private setupWebTracking(): boolean {
    try {
      if (typeof window === 'undefined' || !window.addEventListener) {
        return false;
      }

      const handler = (event: PromiseRejectionEvent) => {
        event.preventDefault(); // Prevent browser's default handling
        const error =
          event.reason instanceof Error
            ? event.reason
            : new Error(String(event.reason));
        this.onRejectionHandler(error);
      };

      window.addEventListener('unhandledrejection', handler);
      this.cleanupFunctions.push(() => {
        window.removeEventListener('unhandledrejection', handler);
      });

      return true;
    } catch (error) {
      return false;
    }
  }

  /**
   * Method 2: Hermes Engine - Use HermesInternal API
   */
  private setupHermesTracking(): boolean {
    try {
      // Check if Hermes engine is available
      const hermesInternal = (global as any).HermesInternal;
      if (
        hermesInternal &&
        typeof hermesInternal.enablePromiseRejectionTracker === 'function'
      ) {
        hermesInternal.enablePromiseRejectionTracker({
          allRejections: true, // Track both handled and unhandled rejections
          onUnhandled: (id: number, error: any) => {
            this.onRejectionHandler(error);
          },
          onHandled: (id: number) => {
            // Promise was handled later - could be used for analytics in the future
          },
        });

        // Cleanup function (if available)
        this.cleanupFunctions.push(() => {
          // Hermes doesn't provide a disable method, but we can set a no-op handler
          if (hermesInternal?.enablePromiseRejectionTracker) {
            hermesInternal.enablePromiseRejectionTracker({
              allRejections: false,
              onUnhandled: () => {},
              onHandled: () => {},
            });
          }
        });

        return true;
      }
      return false;
    } catch (error) {
      return false;
    }
  }

  /**
   * Method 3: React Native Internal - Use promise/setimmediate/rejection-tracking for JSC
   */
  private setupReactNativeTracking(): boolean {
    try {
      // Try to access React Native's internal promise tracking module
      const tracking = require('promise/setimmediate/rejection-tracking');

      if (tracking && typeof tracking.enable === 'function') {
        tracking.enable({
          allRejections: true,
          onUnhandled: (id: number, error: any) => {
            this.onRejectionHandler(error);
          },
          onHandled: (id: number) => {
            // Promise was handled later - tracking for potential analytics
          },
        });

        this.cleanupFunctions.push(() => {
          // Try to disable tracking if possible
          if (tracking.disable && typeof tracking.disable === 'function') {
            tracking.disable();
          }
        });

        return true;
      }
      return false;
    } catch (error) {
      return false;
    }
  }

  /**
   * Clean up all tracking mechanisms
   */
  public cleanup(): void {
    // Clean up all tracking mechanisms and restore original state
    this.cleanupFunctions.forEach((cleanup) => {
      try {
        cleanup();
      } catch (error) {
        // Silently handle cleanup errors to avoid disrupting app shutdown
      }
    });
    this.cleanupFunctions = [];
  }
}

/**
 * Helper function to detect JavaScript engine (simplified for Web/Hermes/JSC)
 */
export function detectJavaScriptEngine(): 'web' | 'hermes' | 'jsc' {
  if (Platform.OS === 'web') {
    return 'web';
  }

  if (typeof (global as any).HermesInternal !== 'undefined') {
    return 'hermes';
  }

  // For non-Hermes engines on mobile, assume JSC (most common)
  return 'jsc';
}

/**
 * Public API for easy integration
 */
/**
 * Sets up advanced promise rejection tracking with automatic engine detection
 * @param onRejection Callback function to handle unhandled promise rejections
 * @returns AdvancedPromiseRejectionTracker instance for cleanup and testing
 */
export function setupAdvancedPromiseRejectionTracking(
  onRejection: PromiseRejectionHandler
): AdvancedPromiseRejectionTracker {
  const tracker = new AdvancedPromiseRejectionTracker(onRejection);
  tracker.setup();
  return tracker;
}
