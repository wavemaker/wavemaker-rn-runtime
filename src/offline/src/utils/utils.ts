export const WM_LOCAL_OFFLINE_CALL = 'WM_LOCAL_OFFLINE_CALL';

export const escapeName = (name: string) => {
    if (name) {
        name = name.replace(/"/g, '""');
        return '"' + name + '"';
    }
    return name;
};
