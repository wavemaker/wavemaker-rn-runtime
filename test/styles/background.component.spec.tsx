import React from 'react';
import { Image } from 'react-native';
import { cleanup, render, waitFor } from '@testing-library/react-native';
import { BackgroundComponent } from '@wavemaker/app-rn-runtime/styles/background.component';
import { AssetProvider } from '@wavemaker/app-rn-runtime/core/asset.provider';
import { isFullPathUrl } from '@wavemaker/app-rn-runtime/core/utils';

const timer = (time = 200) =>
  new Promise((resolve: any, reject) => {
    setTimeout(() => resolve(), time);
  });

const hexToArgbValue = (hex: string) => {
  let hexCode = hex.substring(1);
  if (hexCode.length === 3) {
    hexCode = '' + hex[0] + hex[0] + hex[1] + hex[1] + hex[2] + hex[2];
  }
  hexCode = '0xff' + hexCode;

  return parseInt(hexCode, 16);
};

const formateDecimalValue = (valueInStr: string) => {
  return parseFloat(valueInStr) / 100;
}

const renderComponent = (props = {}) => {
  const defaultProps = {
    size: '100% 100%',
  };

  return render(
    <AssetProvider value={(path) => path}>
      <BackgroundComponent {...defaultProps} {...props} />
    </AssetProvider>
  );
};

// jest.mock('@wavemaker/app-rn-runtime/core/utils', () => {
//   return {
//     isFullPathUrl: jest.fn(),
//   };
// });

jest.mock('@wavemaker/app-rn-runtime/core/imageSizeEstimator', () => ({
  getSize: jest.fn((uri, callback) => {
    callback(100, 200); // mock image dimensions
  }),
}));

describe('Background Component', () => {
  afterEach(() => {
    jest.clearAllMocks();
    cleanup();
  });

  test('should render gradient when image prop contains linear-gradient', () => {
    const tree = renderComponent({
      image: 'linear-gradient(45deg, #4c669f, #3b5998)',
    });

    expect(tree.getByTestId('wm-expo-linear-gradient').props.colors).toEqual([
      hexToArgbValue('#4c669f'),
      hexToArgbValue('#3b5998'),
    ]);
    expect(tree).toMatchSnapshot();
  });

  test('should render gradient when image prop contains linear-gradient with decimal values in gradient color', () => {
    const tree = renderComponent({
      image: 'linear-gradient(68.44deg, #EEEEEE 10%, #A61527 41.98%, #C32033 73.75%, #B41227 100%)',
    });

    expect(tree.getByTestId('wm-expo-linear-gradient').props.colors).toEqual([
      hexToArgbValue('#EEEEEE'),
      hexToArgbValue('#A61527'),
      hexToArgbValue('#C32033'),
      hexToArgbValue('#B41227'),
    ]);
    expect(tree.getByTestId('wm-expo-linear-gradient').props.locations).toEqual([
      formateDecimalValue('10%'),
      formateDecimalValue('41.98%'),
      formateDecimalValue('73.75%'),
      formateDecimalValue('100%')
    ])
    expect(tree).toMatchSnapshot();
  });

  test('source should not have "url" prefix of the image string', async () => {
    const props = {
      image: 'url(http://example.com/image.jpg)',
    };

    const tree = renderComponent(props);
    await timer();

    await waitFor(() => {
      expect(tree.toJSON()).not.toBeNull();
      expect(tree).toMatchSnapshot();
      expect(tree.UNSAFE_getByType(Image).props).toMatchObject({
        resizeMode: 'repeat',
        source: { uri: 'http://example.com/image.jpg' },
      });
    });
  });

  test('should apply image natural width and height when size is falsy', async () => {
    const props = {
      image: 'url(http://example.com/image.jpg)',
      size: null,
    };

    const tree = renderComponent(props);
    await timer();

    await waitFor(() => {
      expect(tree.toJSON()).not.toBeNull();
      expect(tree).toMatchSnapshot();
      expect(tree.UNSAFE_getByType(Image).props).toMatchObject({
        resizeMode: 'repeat',
        source: { uri: 'http://example.com/image.jpg' },
      });

      // value is mocked at the top
      expect(tree.UNSAFE_getByType(Image).props.style).toEqual(
        expect.arrayContaining([
          {
            minHeight: 200,
            minWidth: 100,
          },
        ])
      );
    });
  });

  test('should apply resizeMode correctly', async () => {
    const props = {
      image: 'url(http://example.com/image.jpg)',
      resizeMode: 'center',
    };

    const tree = renderComponent(props);
    await timer();

    await waitFor(() => {
      expect(tree.toJSON()).not.toBeNull();
      expect(tree).toMatchSnapshot();
      expect(tree.UNSAFE_getByType(Image).props).toMatchObject({
        resizeMode: 'center',
        source: { uri: 'http://example.com/image.jpg' },
      });
    });
  });

  xit('should apply position center correctly', async () => {
    const props = {
      image: 'url(http://example.com/image.jpg)',
      position: 'center',
    };

    const tree = renderComponent(props);

    await waitFor(() => {
      expect(tree.toJSON()).not.toBeNull();
      expect(tree.UNSAFE_getByType(Image).props).toMatchObject({
        resizeMode: 'center',
        source: { uri: 'http://example.com/image.jpg' },
      });
      expect(tree.UNSAFE_getByType(Image).parent?.props.style).toEqual(
        expect.arrayContaining([
          {
            alignItems: 'center',
            justifyContent: 'center',
          },
        ])
      );
      expect(tree).toMatchSnapshot();
    });
  });

  test('should apply resizeMode as repeat', async () => {
    const props = {
      image: 'url(http://example.com/image.jpg)',
    };

    const tree = renderComponent(props);

    await waitFor(() => {
      expect(tree.toJSON()).not.toBeNull();
      expect(tree.UNSAFE_getByType(Image).props).toMatchObject({
        resizeMode: 'repeat',
        source: { uri: 'http://example.com/image.jpg' },
      });
      expect(tree).toMatchSnapshot();
    });
  });

  test('should render with image file present in project directory', async () => {
    const props = {
      image: 'resources/images/american-impressionism.png',
    };

    const tree = renderComponent(props);
    await timer();

    await waitFor(() => {
      expect(tree.toJSON()).not.toBeNull();
      expect(tree).toMatchSnapshot();
      expect(tree.UNSAFE_getByType(Image).props).toMatchObject({
        resizeMode: 'repeat',
        source: 'resources/images/american-impressionism.png',
      });
    });
  });

  test('should update image source if props change', async () => {
    const tree = renderComponent({
      image: 'example.com/image.jpg',
      size: '80% 50%',
    });

    await waitFor(() => {
      expect(tree.toJSON()).not.toBeNull();
      expect(tree.UNSAFE_getByType(Image).props).toMatchObject({
        resizeMode: 'repeat',
        source: 'example.com/image.jpg',
      });
      expect(tree.UNSAFE_getByType(Image).parent?.props.style).toEqual(
        expect.arrayContaining([
          {
            height: '50%',
            width: '80%',
          },
        ])
      );
      expect(tree).toMatchSnapshot();
    });

    tree.rerender(
      <AssetProvider value={(path) => path}>
        <BackgroundComponent
          image="example.com/new-image.jpg"
          size="20% 100%"
        />
      </AssetProvider>
    );

    await waitFor(() => {
      expect(tree.toJSON()).not.toBeNull();
      expect(tree.UNSAFE_getByType(Image).props).toMatchObject({
        resizeMode: 'repeat',
        source: 'example.com/new-image.jpg',
      });
      expect(tree.UNSAFE_getByType(Image).parent?.props.style).toEqual(
        expect.arrayContaining([
          {
            height: '100%',
            width: '20%',
          },
        ])
      );
      expect(tree).toMatchSnapshot();
    });
  });

  test('should render image when valid image full path is provided', async () => {
    const tree = renderComponent({
      image: 'http://example.com/image.jpg',
    });

    await waitFor(() => {
      expect(tree.toJSON()).not.toBeNull();
      expect(tree.UNSAFE_getByType(Image).props.source).toMatchObject({
        uri: 'http://example.com/image.jpg',
      });
      expect(tree).toMatchSnapshot();
    });
  });

  test('should set resizeMode based on size prop as "contain"', async () => {
    const tree = renderComponent({
      image: 'http://example.com/image.jpg',
      size: 'contain',
    });

    await waitFor(() => {
      expect(tree.toJSON()).not.toBeNull();
      expect(tree.UNSAFE_getByType(Image).props).toMatchObject({
        source: { uri: 'http://example.com/image.jpg' },
        resizeMode: 'contain',
      });
      expect(tree).toMatchSnapshot();
    });
  });

  xit('should render when position has two values', async () => {
    const tree = renderComponent({
      image: 'http://example.com/image.jpg',
      position: '(top, left)',
    });

    await waitFor(() => {
      expect(tree.toJSON()).not.toBeNull();
      expect(tree.UNSAFE_getByType(Image).props).toMatchObject({
        source: { uri: 'http://example.com/image.jpg' },
      });
      expect(tree.UNSAFE_getByType(Image).parent?.props.style).toEqual(
        expect.arrayContaining([
          {
            left: 0,
            top: 0,
          },
        ])
      );
      expect(tree).toMatchSnapshot();
    });

    tree.rerender(
      <AssetProvider value={(path) => path}>
        <BackgroundComponent
          image="http://example.com/image.jpg"
          position="(top, right)"
        />
      </AssetProvider>
    );

    await waitFor(() => {
      expect(tree.toJSON()).not.toBeNull();
      expect(tree.UNSAFE_getByType(Image).props).toMatchObject({
        source: { uri: 'http://example.com/image.jpg' },
      });
      expect(tree.UNSAFE_getByType(Image).parent?.props.style).toEqual(
        expect.arrayContaining([
          {
            right: 0,
            top: 0,
          },
        ])
      );
      expect(tree).toMatchSnapshot();
    });

    tree.rerender(
      <AssetProvider value={(path) => path}>
        <BackgroundComponent
          image="http://example.com/image.jpg"
          position={'(bottom, right)'}
        />
      </AssetProvider>
    );

    await waitFor(() => {
      expect(tree.toJSON()).not.toBeNull();
      expect(tree.UNSAFE_getByType(Image).props).toMatchObject({
        source: { uri: 'http://example.com/image.jpg' },
      });
      expect(tree.UNSAFE_getByType(Image).parent?.props.style).toEqual(
        expect.arrayContaining([
          {
            right: 0,
            bottom: 0,
          },
        ])
      );
      expect(tree).toMatchSnapshot();
    });

    tree.rerender(
      <AssetProvider value={(path) => path}>
        <BackgroundComponent
          image="http://example.com/image.jpg"
          position={'(bottom, left)'}
        />
      </AssetProvider>
    );

    await waitFor(() => {
      expect(tree.toJSON()).not.toBeNull();
      expect(tree.UNSAFE_getByType(Image).props).toMatchObject({
        source: { uri: 'http://example.com/image.jpg' },
      });
      expect(tree.UNSAFE_getByType(Image).parent?.props.style).toEqual(
        expect.arrayContaining([
          {
            left: 0,
            bottom: 0,
          },
        ])
      );
      expect(tree).toMatchSnapshot();
    });
  });

  test('should render with different repeat props', async () => {
    const tree = renderComponent({
      image: 'http://example.com/image.jpg',
      repeat: 'no-repeat',
      size: null,
    });

    await waitFor(() => {
      expect(tree.toJSON()).not.toBeNull();
      expect(tree.UNSAFE_getByType(Image).props).toMatchObject({
        source: { uri: 'http://example.com/image.jpg' },
      });
      expect(tree.UNSAFE_getByType(Image).parent?.props.style).toEqual(
        expect.arrayContaining([
          {
            height: 200,
            width: 100,
          },
        ])
      );
      expect(tree).toMatchSnapshot();
    });

    tree.rerender(
      <AssetProvider value={(path) => path}>
        <BackgroundComponent
          image="http://example.com/image.jpg"
          repeat="repeat-x"
          size={null}
        />
      </AssetProvider>
    );

    await waitFor(() => {
      expect(tree.toJSON()).not.toBeNull();
      expect(tree.UNSAFE_getByType(Image).props).toMatchObject({
        source: { uri: 'http://example.com/image.jpg' },
        resizeMode: 'repeat',
      });
      expect(tree.UNSAFE_getByType(Image).parent?.props.style).toEqual(
        expect.arrayContaining([
          {
            height: 200,
            width: '100%',
          },
        ])
      );
      expect(tree).toMatchSnapshot();
    });

    tree.rerender(
      <AssetProvider value={(path) => path}>
        <BackgroundComponent
          image="http://example.com/image.jpg"
          repeat="repeat-y"
          size={null}
        />
      </AssetProvider>
    );

    await waitFor(() => {
      expect(tree.toJSON()).not.toBeNull();
      expect(tree.UNSAFE_getByType(Image).props).toMatchObject({
        source: { uri: 'http://example.com/image.jpg' },
      });
      expect(tree.UNSAFE_getByType(Image).parent?.props.style).toEqual(
        expect.arrayContaining([
          {
            width: 100,
            height: '100%',
          },
        ])
      );
      expect(tree).toMatchSnapshot();
    });
  });
});
