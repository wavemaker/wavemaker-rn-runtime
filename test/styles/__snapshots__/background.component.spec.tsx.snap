// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Background Component should apply image natural width and height when size is falsy 1`] = `
<View
  style={
    [
      {
        "borderWidth": 0,
        "overflow": "hidden",
      },
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "alignItems": undefined,
          "flexDirection": "row",
          "justifyContent": undefined,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "overflow": "hidden",
            "position": "absolute",
          },
          undefined,
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
        ]
      }
    >
      <Image
        resizeMode="repeat"
        source={
          {
            "uri": "http://example.com/image.jpg",
          }
        }
        style={
          [
            {
              "height": "100%",
              "width": "100%",
            },
            {
              "minHeight": 200,
              "minWidth": 100,
            },
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`Background Component should apply resizeMode as repeat 1`] = `
<View
  style={
    [
      {
        "borderWidth": 0,
        "overflow": "hidden",
      },
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "alignItems": undefined,
          "flexDirection": "row",
          "justifyContent": undefined,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "overflow": "hidden",
            "position": "absolute",
          },
          undefined,
          {
            "height": "100%",
            "width": "100%",
          },
        ]
      }
    >
      <Image
        resizeMode="repeat"
        source={
          {
            "uri": "http://example.com/image.jpg",
          }
        }
        style={
          [
            {
              "height": "100%",
              "width": "100%",
            },
            null,
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`Background Component should apply resizeMode correctly 1`] = `
<View
  style={
    [
      {
        "borderWidth": 0,
        "overflow": "hidden",
      },
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "alignItems": undefined,
          "flexDirection": "row",
          "justifyContent": undefined,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "overflow": "hidden",
            "position": "absolute",
          },
          undefined,
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
        ]
      }
    >
      <Image
        resizeMode="center"
        source={
          {
            "uri": "http://example.com/image.jpg",
          }
        }
        style={
          [
            {
              "height": "100%",
              "width": "100%",
            },
            {
              "minHeight": 200,
              "minWidth": 100,
            },
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`Background Component should render gradient when image prop contains linear-gradient 1`] = `
<View
  style={
    [
      {
        "borderWidth": 0,
        "overflow": "hidden",
      },
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        null,
      ]
    }
  >
    <ViewManagerAdapter_ExpoLinearGradient
      colors={
        [
          4283197087,
          4282079640,
        ]
      }
      endPoint={
        [
          1,
          0,
        ]
      }
      locations={
        [
          null,
          null,
        ]
      }
      startPoint={
        [
          0,
          1,
        ]
      }
      style={
        [
          {
            "height": "100%",
            "width": "100%",
          },
          undefined,
          {
            "borderRadius": undefined,
            "position": "absolute",
          },
        ]
      }
      testID="wm-expo-linear-gradient"
    />
  </View>
</View>
`;

exports[`Background Component should render gradient when image prop contains linear-gradient with decimal values in gradient color 1`] = `
<View
  style={
    [
      {
        "borderWidth": 0,
        "overflow": "hidden",
      },
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        null,
      ]
    }
  >
    <ViewManagerAdapter_ExpoLinearGradient
      colors={
        [
          4293848814,
          4289074471,
          4290977843,
          4289991207,
        ]
      }
      endPoint={
        [
          1,
          0.3,
        ]
      }
      locations={
        [
          0.1,
          0.41979999999999995,
          0.7375,
          1,
        ]
      }
      startPoint={
        [
          0,
          0.7,
        ]
      }
      style={
        [
          {
            "height": "100%",
            "width": "100%",
          },
          undefined,
          {
            "borderRadius": undefined,
            "position": "absolute",
          },
        ]
      }
      testID="wm-expo-linear-gradient"
    />
  </View>
</View>
`;

exports[`Background Component should render image when valid image full path is provided 1`] = `
<View
  style={
    [
      {
        "borderWidth": 0,
        "overflow": "hidden",
      },
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "alignItems": undefined,
          "flexDirection": "row",
          "justifyContent": undefined,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "overflow": "hidden",
            "position": "absolute",
          },
          undefined,
          {
            "height": "100%",
            "width": "100%",
          },
        ]
      }
    >
      <Image
        resizeMode="repeat"
        source={
          {
            "uri": "http://example.com/image.jpg",
          }
        }
        style={
          [
            {
              "height": "100%",
              "width": "100%",
            },
            null,
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`Background Component should render with different repeat props 1`] = `
<View
  style={
    [
      {
        "borderWidth": 0,
        "overflow": "hidden",
      },
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "alignItems": undefined,
          "flexDirection": "row",
          "justifyContent": undefined,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "overflow": "hidden",
            "position": "absolute",
          },
          undefined,
          {
            "height": 200,
            "width": 100,
          },
        ]
      }
    >
      <Image
        resizeMode="cover"
        source={
          {
            "uri": "http://example.com/image.jpg",
          }
        }
        style={
          [
            {
              "height": "100%",
              "width": "100%",
            },
            null,
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`Background Component should render with different repeat props 2`] = `
<View
  style={
    [
      {
        "borderWidth": 0,
        "overflow": "hidden",
      },
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "alignItems": undefined,
          "flexDirection": "row",
          "justifyContent": undefined,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "overflow": "hidden",
            "position": "absolute",
          },
          undefined,
          {
            "height": 200,
            "width": "100%",
          },
        ]
      }
    >
      <Image
        resizeMode="repeat"
        source={
          {
            "uri": "http://example.com/image.jpg",
          }
        }
        style={
          [
            {
              "height": "100%",
              "width": "100%",
            },
            null,
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`Background Component should render with different repeat props 3`] = `
<View
  style={
    [
      {
        "borderWidth": 0,
        "overflow": "hidden",
      },
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "alignItems": undefined,
          "flexDirection": "row",
          "justifyContent": undefined,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "overflow": "hidden",
            "position": "absolute",
          },
          undefined,
          {
            "height": "100%",
            "width": 100,
          },
        ]
      }
    >
      <Image
        resizeMode="repeat"
        source={
          {
            "uri": "http://example.com/image.jpg",
          }
        }
        style={
          [
            {
              "height": "100%",
              "width": "100%",
            },
            null,
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`Background Component should render with image file present in project directory 1`] = `
<View
  style={
    [
      {
        "borderWidth": 0,
        "overflow": "hidden",
      },
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "alignItems": undefined,
          "flexDirection": "row",
          "justifyContent": undefined,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "overflow": "hidden",
            "position": "absolute",
          },
          undefined,
          {
            "height": "100%",
            "width": "100%",
          },
        ]
      }
    >
      <Image
        resizeMode="repeat"
        source="resources/images/american-impressionism.png"
        style={
          [
            {
              "height": "100%",
              "width": "100%",
            },
            null,
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`Background Component should set resizeMode based on size prop as "contain" 1`] = `
<View
  style={
    [
      {
        "borderWidth": 0,
        "overflow": "hidden",
      },
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "alignItems": undefined,
          "flexDirection": "row",
          "justifyContent": undefined,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "overflow": "hidden",
            "position": "absolute",
          },
          undefined,
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
        ]
      }
    >
      <Image
        resizeMode="contain"
        source={
          {
            "uri": "http://example.com/image.jpg",
          }
        }
        style={
          [
            {
              "height": "100%",
              "width": "100%",
            },
            null,
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`Background Component should update image source if props change 1`] = `
<View
  style={
    [
      {
        "borderWidth": 0,
        "overflow": "hidden",
      },
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "alignItems": undefined,
          "flexDirection": "row",
          "justifyContent": undefined,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "overflow": "hidden",
            "position": "absolute",
          },
          undefined,
          {
            "height": "50%",
            "width": "80%",
          },
        ]
      }
    >
      <Image
        resizeMode="repeat"
        source="example.com/image.jpg"
        style={
          [
            {
              "height": "100%",
              "width": "100%",
            },
            null,
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`Background Component should update image source if props change 2`] = `
<View
  style={
    [
      {
        "borderWidth": 0,
        "overflow": "hidden",
      },
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "alignItems": undefined,
          "flexDirection": "row",
          "justifyContent": undefined,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "overflow": "hidden",
            "position": "absolute",
          },
          undefined,
          {
            "height": "100%",
            "width": "20%",
          },
        ]
      }
    >
      <Image
        resizeMode="repeat"
        source="example.com/new-image.jpg"
        style={
          [
            {
              "height": "100%",
              "width": "100%",
            },
            null,
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`Background Component source should not have "url" prefix of the image string 1`] = `
<View
  style={
    [
      {
        "borderWidth": 0,
        "overflow": "hidden",
      },
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "alignItems": undefined,
          "flexDirection": "row",
          "justifyContent": undefined,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "overflow": "hidden",
            "position": "absolute",
          },
          undefined,
          {
            "height": "100%",
            "width": "100%",
          },
        ]
      }
    >
      <Image
        resizeMode="repeat"
        source={
          {
            "uri": "http://example.com/image.jpg",
          }
        }
        style={
          [
            {
              "height": "100%",
              "width": "100%",
            },
            null,
          ]
        }
      />
    </View>
  </View>
</View>
`;
